include/revng/ADT/STLExtras.h

```diff
@@ -211,9 +211,14 @@ skipImpl(IteratorType &&From,
 
   std::ptrdiff_t TotalSkippedCount = Front + Back;
   if constexpr (forward_iterator<IteratorType>) {
-    // We cannot compute the assert on the input iterators because it's
-    // going to consume them.
-    revng_assert(std::distance(From, To) >= TotalSkippedCount);
+    // We cannot check on the input iterators because it's going to consume
+    // them.
+
+    if (std::distance(From, To) <= TotalSkippedCount) {
+      // Quietly return an empty range if there are more skips requested than
+      // the total number of elements the input range contains.
+      return llvm::make_range(To, To);
+    }
```

Comment:
Mmmmh, I think the `==` part is superfluous.
For the `<` part, are you sure? It seems like a footgun to me.

At very very least this should be very explicitly stated in doxygen.

Reply:
To be fare, I'm really hesitant about this change in general. Because, as you noticed, it's not 100% obvious to everyone.

On the other hand, I felt really bad typing 
```cpp
if (MyRange.size() > X) for (auto &E : skip_front(MyRange, X)) do_stuff();
```

instead of letting `skip_front` figure it out.

Consider that the entire point of the clause is "If you want to skip more elements than you have, don't iterate over anything (drop out of the loop)", which feels more or less reasonable.

👍 on the comment.

Reply:
I guess it's pretty simple to introduce two versions: one asserting and the other one that skips short ranges.

How does
```cpp
skip_front() // returns nothing (two copies of the same iterator) on fail
safe_skip_front() // asserts
```
sound?

Reply:
Every case I used it for needs such an `if`, be it the ABI one, the one in this PR, or the one in the graph one.

The first example I found:
https://github.com/revng/revng/pull/274/files/1f83642c3f1fa4ead339e80d6c853f7f9b36a340..94ffbfd9c68f93668e404376c9afdc1b3a3a14d5#diff-9fc9a9c46221bd88ccba85350b3e949fd3d4e71c11c862b1bfa17d3e342c3997R160

Maybe it's a coincidence that all the cases I work with are content with their inputs being empty, where there's nothing to do.

Reply:
Actually yeah, that's definitely the way we should do it. Thanks :)

Reply:
Please take a fresh look. The commit is `STLExtras: introduce pairwise skip flavour`